clients/src/main/java/org/apache/kafka/common/header/RecordHeaders.java

```diff
@@ -0,0 +1,207 @@
+            this.headers = new ArrayList<>();
+            Iterator<Header> iterator = headers.iterator();
+            while (iterator.hasNext()) {
+                this.headers.add(iterator.next());
+            }
+        }
+    }
+
+    @Override
+    public Headers add(Header header) throws IllegalStateException {
+        checkClosed();
+        headers.add(header);
+        return this;
+    }
+
+    @Override
+    public Headers remove(String key) throws IllegalStateException {
+        checkClosed();
+        checkKey(key);
+        Iterator<Header> iterator = iterator();
+        while (iterator.hasNext()) {
+            if (iterator.next().key().equals(key)) {
+                iterator.remove();
+            }
+        }
+        return this;
+    }
+
+    @Override
+    public Header lastHeader(String key) {
+        checkKey(key);
+        for (int i = headers.size() - 1; i >= 0; i--) {
+            Header header = headers.get(i);
+            if (header.key().equals(key)) {
+                return header;
+            }
+        }
+        return null;
+    }
+
+    @Override
+    public Iterable<Header> headers(final String key) {
+        checkKey(key);
+        return new Iterable<Header>() {
+            @Override
+            public Iterator<Header> iterator() {
+                return new FilterByKeyIterator(headers.iterator(), key);
+            }
+        };
+    }
+
+    @Override
+    public Iterator<Header> iterator() {
+        return closeAware(headers.iterator());
+    }
+
+    @Override
+    public void close() throws IOException {
+        this.isClosed = true;
+    }
+
+    public Header[] toArray() {
+        return headers.isEmpty() ? EMPTY_HEADERS : headers.toArray(new Header[headers.size()]);
+    }
+    
+    private void checkKey(String key) {
+        if (key == null) {
+            throw new IllegalArgumentException("key cannot be null");
+        }
+    }
+    
+    private void checkClosed() {
+        if (isClosed) {
+            throw new IllegalStateException("RecordHeaders isClosed");
+        }
+    }
+
+    private Iterator<Header> closeAware(final Iterator<Header> original) {
+        return new Iterator<Header>() {
+            @Override
+            public boolean hasNext() {
+                return original.hasNext();
+            }
+
+            public Header next() {
+                return original.next();
+            }
+
+            @Override
+            public void remove() {
+                checkClosed();
+                original.remove();
+            }
+        };
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+
+        RecordHeaders headers1 = (RecordHeaders) o;
+
+        return headers != null ? headers.equals(headers1.headers) : headers1.headers == null;
```

Comment:
So this is to the equality of arraylist equals. I believe therefor we get this for free.

As per Java doc 

Compares the specified object with this list for equality. Returns true if and only if the specified object is also a list, both lists have the same size, and all corresponding pairs of elements in the two lists are equal. (Two elements e1 and e2 are equal if (e1==null ? e2==null : e1.equals(e2)).) In other words, two lists are defined to be equal if they contain the same elements in the same order.